### PR TITLE
Reflected dunder methods for jitclass

### PR DESCRIPTION
**Pull request by louisamand**
Tested the addition of the `__r*__` methods in the `supported_dunder`.
As it worked, I added a small batch of tests and updated the doc.

### CODE REVIEW

**Comment by louisamand** [comment]
I don't have access to a Linux computer, but the problem appears to be a memory leak in the case of `vanilla`and/or `cov_doc` tests.
```
Traceback (most recent call last):
  File "/home/vsts/work/1/s/numba/tests/support.py", line 906, in tearDown
    self.memory_leak_teardown()
  File "/home/vsts/work/1/s/numba/tests/support.py", line 881, in memory_leak_teardown
    self.assert_no_memory_leak()
  File "/home/vsts/work/1/s/numba/tests/support.py", line 890, in assert_no_memory_leak
    self.assertEqual(total_alloc, total_free)
AssertionError: 8 != 2
```

**Comment by guilhermeleobas** [comment]
I think this issue arises because how `__matmul__` is lowered in Numba. It requires SciPy and there's no fallback implementation. Could you skip `__matmul__` test if SciPy is not available? There's a fixture in `numba/tests/support.py` called `skip_unless_scipy` which can be of some help.

**Comment by guilhermeleobas** [comment]
HI @louisamand, thanks for the patch. The flake8 errors are:

```
numba/tests/test_jitclasses.py:1877:1: W293 blank line contains whitespace
numba/tests/test_jitclasses.py:1911:81: E501 line too long (126 > 80 characters)
numba/tests/test_jitclasses.py:1916:13: F841 local variable 'py_ops_defined' is assigned to but never used
numba/tests/test_jitclasses.py:1917:13: F841 local variable 'py_ops_not_defined' is assigned to but never used
numba/tests/test_jitclasses.py:1919:13: F841 local variable 'jit_ops_defined' is assigned to but never used
numba/tests/test_jitclasses.py:1920:13: F841 local variable 'jit_ops_not_defined' is assigned to but never used
numba/tests/test_jitclasses.py:1924:81: E501 line too long (85 > 80 characters)
numba/tests/test_jitclasses.py:1925:81: E501 line too long (87 > 80 characters)
numba/tests/test_jitclasses.py:1927:81: E501 line too long (97 > 80 characters)
numba/tests/test_jitclasses.py:1928:81: E501 line too long (99 > 80 characters)

```

**Comment by jsbryaniv** [comment]
I want to point out that the `__matmul__` dunder itself is the issue. The issue is the way the '@' operator works within numba. For user defined matrix multiplication operators there is no issue.
```
        spec = {"x": types.Array(float64, 2, 'C')}
        @jitclass(spec)
        class MinFail:
            def __init__(self, x):
                self.x = x
            
            def __matmul__(self, other):
                output = np.zeros((self.x.shape[0], other.x.shape[1]))
                for i in range(self.x.shape[0]):
                    for k in range(self.x.shape[1]):
                        for j in range(other.x.shape[1]):
                            output[i, j] += self.x[i, k] * other.x[k, j]
                return output
            
        testvals = float64(np.arange(4).reshape(2, 2))
        a = MinFail(testvals)
        b = MinFail(testvals)
        c = a @ b
        print(c)
```

File: numba/tests/test_jitclasses.py [hunk 1]
```diff
@@ -1858,6 +1858,90 @@ def __init__(self, x):
...
+            jit_ops_defined = JitOperatorsDefined(test_value[0])
+            jit_ops_not_defined = JitNoOperatorsDefined(test_value[1])
+
+            for obj in [
+                py_ops_defined, py_ops_not_defined,
+                jit_ops_defined, jit_ops_not_defined
+            ]:
+                obj.x
+                # I haven't found a good way to remove
+                # "local variable 'XXX' is assigned to but never used"
```
Comment (guilhermeleobas) [inline numba/tests/test_jitclasses.py:1931]: You can ignore a flake8 error with `noqa: F841`

File: numba/tests/test_jitclasses.py [hunk 2]
```diff
@@ -1858,6 +1858,90 @@ def __init__(self, x):
...
+        ]:
+            spec = {"x": test_type}
+            JitOperatorsDefined = jitclass(OperatorsDefined, spec)
+            JitNoOperatorsDefined = jitclass(NoOperatorsDefined, spec)
+
+            py_ops_defined = OperatorsDefined(test_value[0])
+            py_ops_not_defined = NoOperatorsDefined(test_value[1])
+
+            jit_ops_defined = JitOperatorsDefined(test_value[0])
+            jit_ops_not_defined = JitNoOperatorsDefined(test_value[1])
```
Comment (guilhermeleobas) [inline numba/tests/test_jitclasses.py:1923]:
```suggestion
            py_ops_defined = OperatorsDefined(test_value[0])  # noqa: F841
            py_ops_not_defined = NoOperatorsDefined(test_value[1])  # noqa: F841

            jit_ops_defined = JitOperatorsDefined(test_value[0])  # noqa: F841
            jit_ops_not_defined = JitNoOperatorsDefined(test_value[1])  # noqa: F841
```

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]
PR look goods. Just one small change from my side.

**Pull request closed without merging**